File: BaseLibrary/command/handshake/HelloCommand.cs
﻿using System;
using System.Text;
using BaseLibrary.config;
using BaseLibrary.protocol;

namespace BaseLibrary.command.handshake {
    public class HelloCommand : AHandShakeCommand, ACommand.Sendable {
        public static readonly IFactory<ACommand.Sendable, ACommand> FACTORY = new HelloCommandFactory();
        private sealed class HelloCommandFactory : ACommandFactory {
            internal HelloCommandFactory() : base() { }

            public override bool IsDeserializeable(string s) {
                if (s.StartsWith("HELLO "+GameProperties.APP_NAME)) {
                    var protocols = s.Substring(("HELLO "+GameProperties.APP_NAME).Length).Split(' ');
                    if (protocols.Length > 0) {
                        cache.Cached(s, new HelloCommand(protocols));
                        return true;
                    }
                }
                return false;
            }

            public override bool IsTransferable(ACommand c) {
                if (c is HelloCommand) {
                    var c1 = (HelloCommand)c;
                    cache.Cached(c, new HelloCommand(c1.SUPPORTED_PROTOCOLS));
                    return true;
                }
                return false;
            }
        }

        public string Serialize() {
            var text = new StringBuilder("HELLO " + GameProperties.APP_NAME);
            foreach (var s in SUPPORTED_PROTOCOLS) {
                text.Append(" " + s.TrimEnd());
            }
            return text.ToString();
        }

        public string[] SUPPORTED_PROTOCOLS { get; private set; }

        public HelloCommand(string[] supportedProtocols)
            : base() {
            Array.Sort(supportedProtocols);
            SUPPORTED_PROTOCOLS = supportedProtocols;
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/command/handshake/HelloCommand.java

package BaseLibrary.communication.command.handshake;

import java.util.Arrays;

import BaseLibrary.communication.command.ACommand;
import BaseLibrary.communication.command.ACommandFactory;
import BaseLibrary.communication.protocol.AProtocol;
import BaseLibrary.communication.protocol.IFactory;
import BaseLibrary.config.GameProperties;

public class HelloCommand extends AHandShakeCommand {
	private static final String NAME = "HELLO";

	static {
		AProtocol.registerForDeserialize(NAME, HelloCommand.class);
	}

	@Override
	public String getCommandName() {
		return NAME;
	}

	public static final IFactory<ACommand, ACommand> FACTORY = new HelloCommandFactory();

	private final static class HelloCommandFactory extends ACommandFactory {
		private HelloCommandFactory() {
		}

		@Override
		public boolean isDeserializeable(String s, AProtocol protocol) {
			if (s.startsWith("HELLO " + GameProperties.APP_NAME)) {
				String[] protocols = s.substring(("HELLO " + GameProperties.APP_NAME).length()).split(" ");
				if (protocols.length > 0) {
					cache.cached(s, new HelloCommand(protocols));
					return true;
				}
			}
			return false;
		}

		@Override
		public boolean isTransferable(ACommand c) {
			if (c instanceof HelloCommand) {
				HelloCommand c1 = (HelloCommand) c;
				cache.cached(c, new HelloCommand(c1.SUPPORTED_PROTOCOLS));
				return true;
			}
			return false;
		}
	}

	public final String[] SUPPORTED_PROTOCOLS;

	public HelloCommand(String[] supportedProtocols) {
		Arrays.sort(supportedProtocols);
		SUPPORTED_PROTOCOLS = supportedProtocols;
	}
	

	@Override
	public String serialize(AProtocol p) {
		StringBuilder text = new StringBuilder("HELLO " + GameProperties.APP_NAME);
		for (String s : SUPPORTED_PROTOCOLS) {
			text.append(" " + s.trim());
		}
		return text.toString();
	}
}
